tools/lcdtools/e2fsck-lcd/e2fsck-lcd.8
.\" $Id: e2fsck-lcd.8 181 2005-05-07 06:51:15Z pdh $
.\"
.\" Copyright (C) 2004, 2005  Martin Michlmayr <tbm@cyrius.com>
.\" This manual is freely distributable under the terms of the GPL.
.\" It was originally written for Debian GNU/Linux (but may be used
.\" by others).
.\"

.TH E2FSCK\-LCD 8 "May 2005"

.SH NAME
e2fsck\-lcd \- output file system check progress information on the Cobalt LCD

.SH SYNOPSIS
.B e2fsck\-lcd

.SH DESCRIPTION
.PP
.B e2fsck\-lcd
is a program which outputs file system check progress information on the
Cobalt LCD.  It will call the real
.B fsck.ext2
and
.B fsck.ext3
programs to perform the file checking functions and output progress
information on the Cobalt LCD.

If called as
.B e2fsck\-lcd
the program will pass command line arguments to
.B fsck.ext2
to perform the file system check. If called by any other name it will
use the called name with
.B \.distrib
appended as the name of the file system check program.

.SH AUTHOR
.PP
Peter Horton <pdh@colonel\-panic.org>



CHANGES
# $Id: CHANGES 195 2006-08-10 22:03:35Z pdh $

CoLo 1.22 (2006-08-10)
----------------------

*	Fix build failure on GCC 4.1.

*	Tweaked INSTALL.
	[Martin Michlmayr]

*	Fixed 'netcon' to not cause an assert in UDP code.

*	Added support for CF cards in CF/IDE adapters (not tested).

CoLo 1.21 (2006-01-22)
----------------------

*	Merged "screen" and "minicom" keymaps.

*	Added "delete word" to line editor ('<ctrl>w').

*	Added 'relocate' command which relocates the initrd image to follow the
	kernel image. Required to use initrd images with current 2.6.x kernels.

*	'mount' command can now be aborted.

CoLo 1.20 (2006-01-16)
----------------------

*	Fix menu cancel/abort/error returns.

*	Update NV format to include keymap.

*	Added keymap for "screen".

*	Wait for network interface transmit ring to drain when disabling network
	interface. This should improve net console output.

*	Added header file dependency checking to stage1, stage2 and chain.

CoLo 1.19 (2005-04-03)
----------------------

*	Fixed stupid bug in ordering of LCD initialisation. This would cause
	installs of 1.17 or 1.18 which weren't an upgrade from 1.16 to crash
	early in stage2.

*	Added "README.tools" to tools directory.

*	Renamed 'md5rom' to 'copy-rom' and added the ability to save a copy of the
	ROM to a file.

*	Moved PCI timeout / retry counter initialisation to PCI initialisation
	code. We now use a retry count of 255 rather than retrying forever. The
	kernel needs a similiar fix.

*	Improved PCI bus scan.

CoLo 1.18 (2005-03-22)
----------------------

*	Added UDP network console support. This adds a 'netcon' command and
	associated variables, and a new boot menu item "Network Shell".

*	Added "ncon" to tools directory. Very simple net console client.

*	PCI serial support switch moved to NV flag 7.

CoLo 1.17 (2005-02-20)
----------------------

*	DHCP client now broadcasts DHCPREQUEST packets. It also includes the
	"server identifier" option from the server in the request. Fixed bug where
	we weren't appending an "end options" tag to the options. [RFC2131]

*	Dropped "no matching socket" diagnostic message.

*	Simplified display of memory size on boot.

*	Added "Cancel" option to 'paneld' menus.

*	Added horizontal scrolling menus to 'paneld' (Controlled by the '-h'
	option).

*	Added horizontal scrolling menus to shell. Controlled by an NV flag (see
	'nvflags' command in README.shell).

*	Reset menu timeout when a button is pressed.

*	Removed ARP diagnostic messages.

CoLo 1.16 (2004-12-16)
----------------------

*	Force menu choice of console on/off on upgrade from earlier versions.

*	Fix 'flash-tool' and 'flash' command to work with TMS29F040 Flash.

CoLo 1.15 (2004-11-28)
----------------------

*	Added link detection to network driver.

*	Added support for 21041 network controller (2700WG units).

*	Added clock speed detection (stage2 only) to ensure timeouts and delays are
	consistent regardless of CPU speed.

*	Added new command 'sleep'.

CoLo 1.14 (2004-10-26)
----------------------

*	Cleaned up build system and added cross compile support.
	[cross compile suggestion / gcc-3.4 link fix - Sebastien Vajda]
	[compiler flag suggestions - Thiemo Seufer]

*	Added new example boot script 'menu.colo' showing new features.

*	Variable expansion in the shell is now carried out before argument
	splitting.  This means variables can now span arguments e.g.

		var message "'first' 'second'"
		lcd {message}

*	Reworked the LCD menu command. The 'menu' command has been removed and
	replaced with a new command, 'select'. The new command justs sets a
	variable to the index of the selected item which can then be used with
	'goto'.

*	Added 'noop' command, which does nothing. When combined with the 'onfail'
	command it can be used for checking the existence of variables.

*	Added script commands 'goto', 'onfail', 'exit' and 'abort'.

*	Added positional marks in scripts using the '@' character.

*	Scripts can now be nested.

*	The 'console-speed' variable is now only available when the serial port is
	enabled.

*	Fixed up support for booting 64-bit kernels (32-bit support was reorganised
	too). 32-bit kernels must be linked for KSEG0 or KSEG1, 64-bit kernels for
	CKSEG0, CKSEG1 or XKPHYS.

*	Ensure we read from Galileo PCI RETRY/TIMEOUT register before writing to
	it. A comment in the kernel source indicates this might be necessary to
	stop Galileo locking up on early units.

*	Added a '-n' command line option to 'putlcd' which allows the user to write
	to a single line of the LCD without affecting the other line.

*	Added big fat warning to 'flash-tool' (just in case).

CoLo 1.13 (2004-08-30)
----------------------

*	Speeded up raw LCD access in LCD tools library.

*	Added 'on' and 'off' options to the 'serial' command.

*	Added 'colo-perm' and 'md5rom' to build.

*	Added missing include to paneld.c.

CoLo 1.12 (2004-08-20)
----------------------

*	The LCD tools 'putlcd', 'paneld' and 'e2fsck-lcd' are now built around a
	common library. The library has support for using direct access to the LCD
	via /dev/mem (the default) or driver access via the Cobalt LCD driver.

	To make the library use the Cobalt LCD driver you need to export an
	environment variable LCD_DEVICE with the name of the device. If LCD_DEVICE
	is an empty string the library will use the default device name, /dev/lcd.

*	Added a '-d' command line option to 'paneld' which makes it fork itself
	into the background on startup.

*	Added 'menu' boot shell command for implementing LCD menus in scripts.
	This feature makes it possible to choose from a list of multiple
	kernels/configurations to boot (see example in README.shell).

*	Fixed a couple of off by one errors in the 'boot' command.

*	Added 'console disable' to the non-volatile flags. This defers initialising
	the serial port until the boot shell is entered and should make it possible
	to use CoLo on units which don't have a serial port. This flag will be set

tools/lcdtools/e2fsck-lcd/e2fsck-lcd.c
/*
 * (C) P.Horton 2004,2005,2006
 *
 * $Id: e2fsck-lcd.c 186 2006-01-17 23:03:58Z pdh $
 *
 * This code is covered by the GNU General Public License. For details see the file "COPYING".
 */

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <errno.h>
#include <assert.h>
#include <sys/wait.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/param.h>

#include "liblcd.h"

#define APP_NAME						"e2fsck-lcd"
#define EXTENSION						".distrib"
#define CHILD_DEFAULT				"fsck.ext2"
#define ERROR_EXIT					8						/* e2fsck - operational error */

#define LCD_WIDTH						16

static const char spaces[] = { [0 ... (LCD_WIDTH - 1)] = ' ' };
static unsigned thumb;
static int lcd;

const char *getapp(void)
{
	return APP_NAME;
}

/*
 * initialise output
 */
static void disp_init(const char *devn)
{
	lcd = lcd_open(NULL);
	if(!lcd)
		return;

	lcd_clear();

	if(devn) {

		lcd_curs_move(0, 0);
		lcd_text(devn, LCD_WIDTH - 7);
	}

	lcd_curs_move(0, LCD_WIDTH - 4);
	lcd_text("0.0%", -1);

	lcd_curs_move(1, 0);
	lcd_text("[>", -1);
	lcd_curs_move(1, LCD_WIDTH - 1);
	lcd_text("]", -1);

	thumb = 0;
}

/*
 * finish output
 */
static void disp_cleanup(void)
{
	usleep(500 * 1000);

	if(lcd)
		lcd_clear();

	fputs("       \r", stdout);
	fflush(stdout);
}

/*
 * update output with new progress information
 */
static void disp_update(int pass, unsigned where, unsigned limit)
{
	static char bar[] = { [0 ... (LCD_WIDTH - 1)] = '=' };
	static char text[16];
	unsigned prog;

	if(pass < 1 || pass > 5)
		return;

	prog = (pass - 1) * 200 + (where * 200 + limit / 2) / limit;

	snprintf(text, sizeof(text), "%3u.%u%%", prog / 10, prog % 10);
	text[sizeof(text) - 1] = '\0';

	if(lcd) {

		lcd_curs_move(0, LCD_WIDTH - 6);
		lcd_text(text, 6);

		prog = ((LCD_WIDTH - 2) * prog + 500) / 1000;
		if(prog > thumb) {
			lcd_curs_move(1, thumb + 1);
			lcd_text(bar, prog - thumb);
			thumb = prog;
			if(thumb < LCD_WIDTH - 2)
				lcd_text(">", -1);
		}
	}

	printf(" %s\r", text);
	fflush(stdout);
}

/*
 * read progress information from pipe
 */
static void monitor(int fd, const char *devn)
{
	static char ibuf[8192];

	unsigned fill, pass, where, limit;
	int cntl, nread, skip, init;
	char pad, *ptr;
	fd_set set;

	cntl = fcntl(fd, F_GETFL);
	if(cntl == -1 || fcntl(fd, F_SETFL, cntl | O_NONBLOCK)) {
		fprintf(stderr, APP_NAME ": failed to set pipe to non-blocking (%s)\n", strerror(errno));
		return;
	}

	if(devn) {
		ptr = strrchr(devn, '/');
		if(ptr)
			devn = ptr + 1;
	}

	FD_ZERO(&set);

	for(init = 0, fill = 0, skip = 0;;) {

		FD_SET(fd, &set);

		if(select(fd + 1, &set, NULL, NULL, NULL) == -1 && errno != EINTR) {
			fprintf(stderr, APP_NAME ": failed waiting on pipe (%s)\n", strerror(errno));
			return;
		}

		nread = read(fd, ibuf + fill, sizeof(ibuf) - fill);
		if(!nread)
			break;

		if(nread == -1) {
			if(errno != EINTR && errno != EAGAIN) {
				fprintf(stderr, APP_NAME ": failed reading from pipe (%s)\n", strerror(errno));
				return;
			}
			continue;
		}

		fill += nread;

		for(;;) {

			ptr = strchr(ibuf, '\n');
			if(!ptr) {

				if(fill == sizeof(ibuf)) {

					if(!skip)
						fputs(APP_NAME ": line too long\n", stderr);

					skip = 1;
					fill = 0;
				}

				break;
			}

			*ptr++ = '\0';

			if(!skip) {

				if(sscanf(ibuf, " %u %u %u %c", &pass, &where, &limit, &pad) == 3) {

					if(!init) {
						disp_init(devn);
						init = 1;
					}

					disp_update(pass, where, limit);

				} else

					fputs(APP_NAME ": line invalid format\n", stderr);
			}

			skip = 0;

			fill -= ptr - ibuf;
			memmove(ibuf, ptr, fill);
		}
	}

	if(init > 0)
		disp_cleanup();
}

int main(int argc, char *argv[])
{
	static char path[PATH_MAX] = CHILD_DEFAULT;

	char *ptr, *end, *desc, *devn, *argp[argc + 1];
	int next, retn, tube[2];
	pid_t child, dead;
	struct stat info;
	unsigned indx;

	/* work out executable to spawn */

	ptr = strrchr(argv[0], '/') + 1;
	if(ptr == (char *) NULL + 1)
		ptr = argv[0];

	if(strcmp(ptr, APP_NAME)) {
		ptr = path + strlen(argv[0]);
		if(ptr + sizeof(EXTENSION) > path + sizeof(path)) {
			fputs(APP_NAME ": executable path too long\n", stderr);
			return ERROR_EXIT;
		}
		strcpy(path, argv[0]);
		strcpy(ptr, EXTENSION);
	}

	argp[0] = path;
	for(indx = 1; indx < argc; ++indx)
		argp[indx] = argv[indx];
	argp[indx] = NULL;

	/* find argument of -C option and device node */

	desc = NULL;
	devn = NULL;
	next = 0;

	for(indx = 0;; ++indx) {

		ptr = argp[indx];
		if(!ptr)
			break;

		if(next) {
			next = 0;
		} else if(*ptr == '-') {
			ptr = strchr(ptr, 'C');
			if(!ptr)
				continue;
			next = !*++ptr;
			if(next)
				continue;
		} else if(strchr(ptr, '/') && !stat(ptr, &info) && S_ISBLK(info.st_mode)) {
			devn = ptr;
			continue;
		}

		if(!desc) {
			strtoul(ptr, &end, 0);
			if(end > ptr && !*end)
				desc = ptr;
		}
	}

	if(desc) {

		/* start e2fsck as child */

		close(3);
		close(4);

		if(pipe(tube))

			fprintf(stderr, APP_NAME ": failed to create pipe (%s)\n", strerror(errno));

		else if(tube[1] < 10) {

			child = fork();
			switch(child) {

				case -1:
					fprintf(stderr, APP_NAME ": failed to fork child (%s)\n", strerror(errno));
					break;

				default:

					close(tube[1]);

					monitor(tube[0], devn);

					close(tube[0]);		/* force SIGPIPE, what else can we do ? */

					for(;;) {

						dead = wait(&retn);
						if(dead == child)
							break;

						if(dead == -1 && errno != EINTR) {
							fprintf(stderr, APP_NAME ": failed waiting for child to exit (%s)\n", strerror(errno));
							return ERROR_EXIT;
						}
					}

					if(WIFSIGNALED(retn)) {
						fprintf(stderr, APP_NAME ": child exited on signal #%d\n", WTERMSIG(retn));
						return ERROR_EXIT;
					}

					return WEXITSTATUS(retn);

				case 0:

					close(tube[0]);

					/* replace -C argument with our descriptor */

					desc[0] = '0' + tube[1];
					desc[1] = '\0';
			}
		}
	}

	execvp(argp[0], argp);

	fprintf(stderr, APP_NAME ": failed to exec \"%s\" (%s)\n", argp[0], strerror(errno));

	return ERROR_EXIT;
}

/* vi:set ts=3 sw=3 cin: */
